### PR TITLE
Rewrite IN in query for remote shards to exclude values that do not belong to the shard

### PR DESCRIPTION
**Pull request by azat**
Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Exclude values that does not belong to the shard from right part of IN section for distributed queries (under `optimize_skip_unused_shards_rewrite_in`, enabled by default, since it still requires `optimize_skip_unused_shards`)

Detailed description / Documentation draft:

### CODE REVIEW

File: src/Interpreters/OptimizeShardingKeyRewriteInVisitor.cpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    const Cluster::ShardInfo & shard_info, const Cluster::SlotToShard & slots)
+{
+    field = convertField(identifier, field, expr);
+    if (field.isNull())
+        return true;
+
+    UInt64 value = field.get<UInt64>();
+    const auto shard_num = slots[value % slots.size()] + 1;
+    bool shard_contains = shard_info.shard_num == shard_num;
+    std::cerr
```
Comment (alexey-milovidov) [inline src/Interpreters/OptimizeShardingKeyRewriteInVisitor.cpp:44]: Remove before merge.

**Comment by azat** [comment]
> [Stress test (address) — Hung check failed](https://clickhouse-test-reports.s3.yandex.net/21511/d7d9638ceb6bd702f34a88ee54f8f83197e90af5/stress_test_(address).html#fail1)
> [Stress test (thread) — Hung check failed](https://clickhouse-test-reports.s3.yandex.net/21511/d7d9638ceb6bd702f34a88ee54f8f83197e90af5/stress_test_(thread).html#fail1)

```
Found hung queries in processlist:
...
query:                alter table lc_empty_part_bug modify column s LowCardinality(String);
```

File: src/Interpreters/ClusterProxy/executeQuery.cpp [hunk 2]
```diff
@@ -117,9 +117,27 @@ void executeQuery(
     else
         throttler = user_level_throttler;
 
+    size_t shards = query_info.cluster->getShardCount();
     for (const auto & shard_info : query_info.cluster->getShardsInfo())
     {
-        stream_factory.createForShard(shard_info, query, query_ast,
+        ASTPtr query_ast_for_shard;
+        if (settings.optimize_skip_unused_shards && settings.optimize_skip_unused_shards_rewrite_in && shards > 1)
```
Comment (azat) [inline src/Interpreters/ClusterProxy/executeQuery.cpp:124]: This should take `optimize_skip_unused_shards_limit` into account (i.e. "was skipping done or not"), that is added in #21512

**Comment by azat** [comment]
@alexey-milovidov any feedback on this one?

**Comment by azat** [comment]
>[Integration tests (asan) — fail: 1, passed: 1176, flaky: 12](https://clickhouse-test-reports.s3.yandex.net/21511/7cc8b3eebcf575eccaa3b77e8ce7445cd74e3c51/integration_tests_(asan).html#fail1)
>[Integration tests (release) — fail: 1, passed: 1165, flaky: 23](https://clickhouse-test-reports.s3.yandex.net/21511/7cc8b3eebcf575eccaa3b77e8ce7445cd74e3c51/integration_tests_(release).html#fail1)
>[Integration tests (thread) — Timeout, fail: 2, passed: 1099, flaky: 22](https://clickhouse-test-reports.s3.yandex.net/21511/7cc8b3eebcf575eccaa3b77e8ce7445cd74e3c51/integration_tests_(thread).html#fail1)

#22183 (and one unrelated failure, that fails not only here)

File: src/Interpreters/ClusterProxy/executeQuery.cpp (hunk 2 above)
Comment (azat) [reply src/Interpreters/ClusterProxy/executeQuery.cpp]: Done

**Comment by azat** [comment]
@alexey-milovidov can you take a look?

**Review by alexey-milovidov — COMMENTED** [review]
Ok. The only concern is that it will be very slow for long lists of values.
Maybe collect all of them into a column and process once?
(once for every shard - at least; collect to a column just once - at most)

**Comment by alexey-milovidov** [comment]
We have one client that will be very happy to enable this option, but they also have very long lists (millions of items).

**Comment by azat** [comment]
>Ok. The only concern is that it will be very slow for long lists of values.

Right now there is `optimize_skip_unused_shards_limit` that will disable this optimization (so at least the server will not spend too much for one query on the initiator with this option).

>Maybe collect all of them into a column and process once?

Good idea, I will take a look at this later.

>We have one client that will be very happy to enable this option, but they also have very long lists (millions of items).

Got it, but note, that it makes sense even for thousands, not just millions.

File: tests/queries/0_stateless/01756_optimize_skip_unused_shards_rewrite_in.sql [hunk 3]
```diff
@@ -0,0 +1,127 @@
+-- NOTE: this test cannot use 'current_database = currentDatabase()',
```
Comment (azat) [inline tests/queries/0_stateless/01756_optimize_skip_unused_shards_rewrite_in.sql:1]: Interesting, this PR does not adds this test to the arcadia skip list, but I guess it had been passed eventually?

**Pull request merged**